src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -187,6 +213,46 @@ await featureFilter.EvaluateAsync(context).ConfigureAwait(false) == targetEvalua
             return enabled;
         }
 
+        private void BindSettings(IFeatureFilterMetadata filter, FeatureFilterEvaluationContext context)
+        {
+            IFilterParametersBinder binder = filter as IFilterParametersBinder;
+
+            if (binder == null)
+            {
+                return;
+            }
+
+            object settings;
+
+            //
+            // Check if settings already bound from configuration
+            ConfigurationCacheItem cacheItem = (ConfigurationCacheItem)_cache.Get(context.FeatureName);
+
+            if (cacheItem == null ||
+                cacheItem.Parameters != context.Parameters)
```

Comment:
It only compares the ref. Which is a breaking change if someone has a custom provider that updates the parameters object instead of changing the ref. Our provider creates a new ref when updating, so it only affects custom providers.

Reply:
I understand it's not optimal. How costly is it to do the value comparison? Alternatively, we don't do the value comparison every time but only when we are about to expire the cache. It could avoid the breaking change.

Reply:
The way things are written now, we'd have to do the check on every IsEnabledAsync() call, because the goal is to catch this before the cache expires. If the cache expires or is about to expire, there's no issue, as we'd just recompute using the same object. 

Which means in the current state- it's only a breaking change for up to 5 seconds. Even then, it doesn't break anything. It simply means their app will update every 5 seconds instead of sooner. Unless they extended the cache without updating their custom provider, which is the fear.

The check would essentially be a checksum on the parameters, which is fairly fast. Which I thought was a good idea, but we'd essentially be re-adding the same delay that this cache was hoping to solve (parsing the json to a typed object). We could cache the checksum, but if that cache was 5 seconds long, it's the exact same scenario we're already in.

Reply:
Agreed, I think that fear has been mitigated. But we should still call it out in our release/readme so no one gets burned by enabling the cache.

Reply:
Thanks for the explanation, Ross. 

I hate to dump another "cache expiration" concept (besides what our config provider already had) on customers. From a customer perspective, it's hard to tell why 5s is a good number. As a customer, I would expect everything just works, and if anything changes, the cache updates itself automatically. This is why I was trying to see how bad it is if we make the value/checksum comparison even upon every IsEnabledAsync call. If it doesn't work, can we figure out something better?

Reply:
Or, if it's the 5 second piece that's confusing, we can just make the switch a boolean.

Reply:
So I ran some numbers. Running IsEnabledAsync on a TargetingFilter 10,000 times each:
No Cache: ~2000ms
Cache: 56ms

I wonder if we can get away with just setting the cache to 1 second or even 0.5 seconds. And then just accepting that "FeatureManagement library has up to a 1 second delay when using a custom provider".

Reply:
Reason being, each instance of turning Params to a typed object in the non-cached world only takes ~0.5ms. 

Which means even if our cache only lives for 100ms, we'd only be adding 0.5% additional time to their total computation time in the worst case. So we get pretty significant gains with small caches.

Reply:
I'm wondering what the perf looks like if we don't have cache expiration at all, but we evict the cache as soon as a change is detected.

If it's our feature definition provider, we do a reference check (as you did already). If it's a custom provider, we do a full value/checksum comparison.

Reply:
Synced with @zhenlan, 

I got some deep equals metrics. For others, this is a comparison of doing a deep equals on Parameters each call vs binding the parameters in each call. Because I fear Deep Equals are too expensive to be helpful in this scenario.

The timing of BindParameters 10000 times is ~600ms
The timing of DeepEquals 10000 times is ~560ms

Rewriting to the cache 10000 times adds ~30ms. which I think should be added to the BindParameters total time.

This isn't enough of a performance boost over Binding to justify doing that check on each IsEnabledAsync. Here's my version of DeepEquals, let me know if you have any improvements that might make it viable:

```
        private bool AreConfigurationsEqual(IConfiguration config1, IConfiguration config2)
        {
            List<string> keys1 = config1.GetChildren().Select(x => x.Key).ToList();
            List<string> keys2 = config2.GetChildren().Select(x => x.Key).ToList();

            if (!keys1.SequenceEqual(keys2))
            {
                return false;
            }

            foreach (string key in keys1)
            {
                IConfigurationSection section1 = config1.GetSection(key);
                IConfigurationSection section2 = config2.GetSection(key);

                if (section1.Value == null)
                {
                    if (!AreConfigurationsEqual(section1, section2))
                    {
                        return false;
                    }
                } else if (section1.Value != section2.Value)
                {
                    return false;
                }
            }

            return true;
        }
```

Reply:
Thanks for making the comparison, @rossgrambo. It doesn't look like it is worth it for custom providers. So, let's scope this optimization to our built-in providers, as Jimmy suggested. We can open it up to custom providers in the next major release (please open an issue to track it).